### PR TITLE
feat: use GHA to deploy this onto a kind server

### PR DESCRIPTION
**Pull request by garethjevans**
Improving the Gtihub Actions so that it tests the deployment of the chart to a number of kind clusters (at different versions).  There are currently some issues with the chart on 1.14/1.15.

This PR also contains minor fixes to ingress rules on 1.19 and above.

Signed-off-by: Gareth Evans <gareth@bryncynfelin.co.uk>

# What this PR does / why we need it

# Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #

# Special notes for your reviewer

# Checklist

- [x] [DCO](https://github.com/jenkinsci/helm-charts/blob/main/CONTRIBUTING.md#sign-off-your-work) signed
- [x] Chart Version bumped
- [x] CHANGELOG.md was updated

### CODE REVIEW

File: .github/workflows/deploy-on-kind.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+    strategy:
+      matrix:
+        kubernetes_version:
+          #  - "kindest/node:v1.20.0" - currently broken due to chart
+          #  - "kindest/node:v1.19.1" - currently broken due to chart
+          - "kindest/node:v1.18.8"
+          - "kindest/node:v1.17.11"
+          - "kindest/node:v1.16.15"
+          #  - "kindest/node:v1.15.12" - does not support startupProbe
```
Comment (timja) [inline .github/workflows/deploy-on-kind.yaml:17]: I don’t think we need 1.15 and 1.14

**Review by timja — APPROVED** [review]

File: .github/workflows/deploy-on-kind.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
+name: Testing
+on: push
```
Comment (timja) [inline .github/workflows/deploy-on-kind.yaml:2]: Does this need to only run if paths in the Jenkins helm chart have run?

i.e we don’t have other charts here atm, but there was plans to add the Jenkins operator at one point

File: .github/workflows/deploy-on-kind.yaml (hunk 1 above)
Comment (garethjevans) [reply .github/workflows/deploy-on-kind.yaml]: i've removed these

File: .github/workflows/deploy-on-kind.yaml (hunk 2 above)
Comment (garethjevans) [reply .github/workflows/deploy-on-kind.yaml]: updated to only run on changes to `charts/jenkins/**`

File: .github/workflows/deploy-on-kind.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        kubernetes_version:
+          - "kindest/node:v1.20.0"
+          - "kindest/node:v1.19.1"
+          - "kindest/node:v1.18.8"
+          - "kindest/node:v1.17.11"
+          - "kindest/node:v1.16.15"
```
Comment (torstenwalter) [inline .github/workflows/deploy-on-kind.yaml:22]: Following https://en.wikipedia.org/wiki/Kubernetes it should be sufficient to test 1.18, 1.19 and 1.20

File: .github/workflows/deploy-on-kind.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        run: helm upgrade --install --wait --set controller.ingress.enabled=true jenkins charts/jenkins
+
+      - name: Test Chart
+        run: helm test jenkins
+
+      - name: Display Pods
+        run: kubectl get pods
+
+      - name: Display Ingress
+        run: kubectl get ing -oyaml
```
Comment (torstenwalter) [inline .github/workflows/deploy-on-kind.yaml:64]: most of these steos are done by chart-testing automatically

File: .github/workflows/deploy-on-kind.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Run chart-testing (lint)
+        run: helm lint charts/jenkins
+
+      - name: Create Kind Cluster
+        uses: helm/kind-action@v1.1.0
+        with:
+          node_image: ${{ matrix.kubernetes_version }}
+
+      - name: Deploy Chart
+        run: helm upgrade --install --wait --set controller.ingress.enabled=true jenkins charts/jenkins
```
Comment (torstenwalter) [inline .github/workflows/deploy-on-kind.yaml:55]: If we want to test controller.ingress then it should be enabled in https://github.com/jenkinsci/helm-charts/blob/main/charts/jenkins/ci/other-values.yaml

File: charts/jenkins/templates/jenkins-controller-ingress.yaml [hunk 6]
```diff
@@ -37,6 +37,7 @@ spec:
             name: {{ template "jenkins.fullname" . }}
             port:
               number: {{ .Values.controller.servicePort }}
+        pathType: ImplementationSpecific
```
Comment (torstenwalter) [inline charts/jenkins/templates/jenkins-controller-ingress.yaml:40]: This is already part of #200

File: .github/workflows/deploy-on-kind.yaml (hunk 3 above)
Comment (garethjevans) [reply .github/workflows/deploy-on-kind.yaml:22]: It depends on what versions of Kubernetes you want to support? for example, GKE stable is still 1.16, (regular is 1.17) - I'd expect other cloud providers to lag at a similar rate.

File: .github/workflows/deploy-on-kind.yaml (hunk 5 above)
Comment (garethjevans) [reply .github/workflows/deploy-on-kind.yaml:55]: I think that testing ingress would be a good idea, this is the area thats currently causing issues for people on 1.19+

File: charts/jenkins/templates/jenkins-controller-ingress.yaml (hunk 6 above)
Comment (garethjevans) [reply charts/jenkins/templates/jenkins-controller-ingress.yaml:40]: it looks like (from my testing), that for `v1beta1`, it defaults to `pathType: ImplementationSpecific`, but for `v1` you need to explicitly set it

File: .github/workflows/deploy-on-kind.yaml (hunk 3 above)
Comment (timja) [reply .github/workflows/deploy-on-kind.yaml:22]: Azure is supporting back till 1.16 until next month:
https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions#aks-kubernetes-release-calendar

Comment (torstenwalter) [reply .github/workflows/deploy-on-kind.yaml:22]: Good arguments to support 1.16
Only downside that ci jobs will take longer.

File: .github/workflows/deploy-on-kind.yaml (hunk 5 above)
Comment (torstenwalter) [reply .github/workflows/deploy-on-kind.yaml:55]: Sure! Every values file in ci folder is used for testing automatically.

File: .github/workflows/deploy-on-kind.yaml (hunk 3 above)
Comment (garethjevans) [reply .github/workflows/deploy-on-kind.yaml:22]: the CI jobs are run in parallel so shouldn't add much time to the overall build.

File: .github/workflows/deploy-on-kind.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
+---
+name: Testing on Kind
```
[the remark this answers is not in the document]
Comment (timja) [reply .github/workflows/deploy-on-kind.yaml:2]: @garethjevans thoughts here ^^

File: .github/workflows/deploy-on-kind.yaml (hunk 3 above)
Comment (torstenwalter) [reply .github/workflows/deploy-on-kind.yaml:22]: Could you integrate the multi cluster matrix in the existing lint-test workflow?

**Review by timja — APPROVED** [review]

**Review by torstenwalter — APPROVED** [review]
@garethjevans Thank you for this! It's a really neat improvement.

@timja I had to change the required status check to check for this:
image

**Pull request merged**